overlord/servicestate/quota_handlers.go

```diff
@@ -0,0 +1,260 @@
+			//    EBUSY, and it isn't clear if the modification is then properly
+			//    in place or not.
+			//
+			// We will already have called daemon-reload at the end of
+			// EnsureSnapServices directly, so handling case 3 is difficult, and
+			// for now we disallow making this sort of change to a quota group,
+			// that logic is handled at a higher level than this function.
+			// Thus the only decision we really have to make is if the slice was
+			// newly written or not, and if it was save it for later
+			if old == "" {
+				grpsToStart = append(grpsToStart, grp)
+			}
+
+		case "service":
+			// in this case, the only way that a service could have been changed
+			// was if it was moved into or out of a slice, in both cases we need
+			// to restart the service
+			sn := app.Snap
+			appsToRestartBySnap[sn] = append(appsToRestartBySnap[sn], app)
+
+			// TODO: what about sockets and timers? activation units just start
+			// the full unit, so as long as the full unit is restarted we should
+			// be okay?
+		}
+	}
+	if err := wrappers.EnsureSnapServices(snapSvcMap, ensureOpts, collectModifiedUnits, meter); err != nil {
+		return err
+	}
+
+	if ensureOpts.Preseeding {
+		return nil
+	}
+
+	// TODO: should this logic move to wrappers in wrappers.RestartGroups()?
+	systemSysd := systemd.New(systemd.SystemMode, meter)
+
+	// now start the slices
+	for _, grp := range grpsToStart {
+		// TODO: what should these timeouts for stopping/restart slices be?
+		if err := systemSysd.Start(grp.SliceFileName()); err != nil {
+			return err
+		}
+	}
+
+	// after starting all the grps that we modified from EnsureSnapServices,
+	// we need to handle the case where a quota was removed, this will only
+	// happen one at a time and can be identified by the grp provided to us
+	// not existing in the state
+	if _, ok := allGrps[grp.Name]; !ok {
+		// stop the quota group, then remove it
+		if !ensureOpts.Preseeding {
+			if err := systemSysd.Stop(grp.SliceFileName(), 5*time.Second); err != nil {
+				logger.Noticef("unable to stop systemd slice while removing group %q: %v", grp.Name, err)
+			}
+		}
+
+		// TODO: this results in a second systemctl daemon-reload which is
+		// undesirable, we should figure out how to do this operation with a
+		// single daemon-reload
+		st.Unlock()
+		err := wrappers.RemoveQuotaGroup(grp, meter)
+		st.Lock()
+		if err != nil {
+			return err
+		}
+	}
+
+	// now restart the services for each snap that was newly moved into a quota
+	// group
+
+	// iterate in a sorted order over the snaps to restart their apps for easy
+	// tests
+	snaps := make([]*snap.Info, 0, len(appsToRestartBySnap))
+	for sn := range appsToRestartBySnap {
+		snaps = append(snaps, sn)
+	}
+
+	sort.Slice(snaps, func(i, j int) bool {
+		return snaps[i].InstanceName() < snaps[j].InstanceName()
+	})
+
+	for _, sn := range snaps {
+		st.Unlock()
+		disabledSvcs, err := wrappers.QueryDisabledServices(sn, meter)
+		st.Lock()
```

Comment:
we will need to think what are the conflict decisions to make about snaps when we introduce the tasks